### PR TITLE
backend/feat/Ambulance Ride Flow

### PR DESCRIPTION
**Pull request by 0utkarsh**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Screenshot of test results

Screenshot 2024-06-26 at 3 00 54 PM

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Status.hs [hunk 1]
```diff
@@ -58,6 +60,9 @@ handler transporterId req = do
         DRide.NEW -> do
           bookingDetails <- SyncRide.fetchBookingDetails ride booking
           driverInfo <- QDI.findById (cast ride.driverId) >>= fromMaybeM DriverInfoNotFound
+          vehicle <- QVeh.findById ride.driverId >>= fromMaybeM (DriverWithoutVehicle ride.driverId.getId)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Status.hs:63]: Why vehicle is required here? If vehicle is being remove from the driver at time on on-status being called? Or it got changed in b/w?

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/OnDemand/Utils/Common.hs [hunk 2]
```diff
@@ -784,6 +790,35 @@ mkTollConfidenceTagGroupV2 tollConfidence' =
         }
     ]
 
+mkVehicleAgeTagGroupV2 :: Maybe Double -> Maybe [Spec.TagGroup]
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/OnDemand/Utils/Common.hs:793]: Use Piyush way of adding tags.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs [hunk 3]
```diff
@@ -439,33 +439,45 @@ buildQuote ::
...
-          nightShiftOverlapChecking = nightShiftOverlapChecking,
-          estimatedDistance = searchRequest.estimatedDistance,
-          estimatedRideDuration = searchRequest.estimatedDuration,
-          timeDiffFromUtc = Nothing,
-          tollCharges = tollCharges,
-          currency = searchRequest.currency,
-          distanceUnit = searchRequest.distanceUnit
-        }
+      isAmbulanceQuote = DTC.isAmbulanceTrip fullFarePolicy.tripCategory
+  (minFareParams, maxFareParams) <- do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:443]: Move to Estimate.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Status.hs [hunk 4]
```diff
@@ -58,6 +60,9 @@ handler transporterId req = do
         DRide.NEW -> do
           bookingDetails <- SyncRide.fetchBookingDetails ride booking
           driverInfo <- QDI.findById (cast ride.driverId) >>= fromMaybeM DriverInfoNotFound
+          vehicle <- QVeh.findById ride.driverId
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Status.hs:63]: Store the age in `RideDetails`.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 5]
```diff
@@ -1152,7 +1153,7 @@ respondQuote (driverId, merchantId, merchantOpCityId) clientId mbBundleVersion m
               distanceUnit = searchReq.distanceUnit,
               ..
             }
-      QFP.updateFareParameters fareParams
+      QFP.updateFareParameters fareParams fareParams.id -- this must not be working?
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:1156]: Remove from here.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 6]
```diff
@@ -1178,6 +1179,39 @@ acceptStaticOfferDriverRequest mbSearchTry driver quoteId reqOfferedValue mercha
   unless (booking.status == DRB.NEW) $ throwError RideRequestAlreadyAccepted
   whenJust mbSearchTry $ \searchTry -> QST.updateStatus DST.COMPLETED searchTry.id
   (ride, _, vehicle) <- initializeRide merchant driver booking Nothing Nothing clientId
+  farePolicy <- getFarePolicyByEstOrQuoteId merchantOpCityId booking.tripCategory booking.vehicleServiceTier booking.area quoteId Nothing (Just (TransactionId (Id booking.transactionId)))
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:1182]: Not needed.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverOnboarding/VehicleRegistrationCertificate.hs [hunk 7]
```diff
@@ -351,6 +351,7 @@ onVerifyRCHandler person rcVerificationResponse mbVehicleCategory mbAirCondition
           permitValidityUpto = convertTextToUTC rcVerificationResponse.permitValidityUpto,
           pucValidityUpto = convertTextToUTC rcVerificationResponse.pucValidityUpto,
           manufacturer = rcVerificationResponse.manufacturer,
+          mYManufacturing = convertTextToDay (rcVerificationResponse.mYManufacturing <> Just "-15"),
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverOnboarding/VehicleRegistrationCertificate.hs:354]: add comment for -15

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/LeaderBoard.hs [hunk 8]
```diff
@@ -22,6 +22,7 @@ import Data.Time.Calendar ()
 import Data.Time.Calendar.OrdinalDate (sundayStartWeek)
 import Domain.Action.UI.Person
 import Domain.Action.UI.Ride.EndRide.Internal as RideEndInt
+import Domain.Types.Common
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/LeaderBoard.hs:25]: remove this.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride.hs [hunk 9]
```diff
@@ -200,6 +200,7 @@ data DriverRideRes = DriverRideRes
     startOdometerReading :: Maybe DRide.OdometerReading,
     endOdometerReading :: Maybe DRide.OdometerReading,
     returnTime :: Maybe UTCTime,
+    vehicleAge :: Maybe Double,
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride.hs:203]: It should be in `Months` and also move it to RideDetails.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FarePolicy/FarePolicyAmbulanceDetails.hs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+import Kernel.Types.Common
+
+newtype FPAmbulanceDetailsD (s :: UsageSafety) = FPAmbulanceDetails
+  { slabs :: NonEmpty (FPAmbulanceDetailsSlabD s)
+  }
+  deriving (Generic, Show, Eq)
+
+data FPAmbulanceDetailsSlabD (s :: UsageSafety) = FPAmbulanceDetailsSlab
+  { id :: Int,
+    vehicleAge :: Double,
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FarePolicy/FarePolicyAmbulanceDetails.hs:28]: Months.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/Quote.hs [hunk 11]
```diff
@@ -35,6 +35,8 @@ data Quote = Quote
     tripCategory :: DTC.TripCategory,
     validTill :: UTCTime,
     estimatedFare :: HighPrecMoney,
+    minEstimatedFare :: Maybe HighPrecMoney,
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/Quote.hs:38]: Remove this.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/SendSearchRequestToDrivers.hs [hunk 12]
```diff
@@ -122,21 +126,52 @@ sendSearchRequestToDrivers tripQuoteDetails searchReq searchTry driverPoolConfig
...
-    buildSearchRequestForDriver tripQuoteDetailsHashMap batchNumber defaultValidTill dpwRes = do
+    buildSearchRequestForDriver tripQuoteDetailsHashMap batchNumber defaultValidTill transporterConfig dpwRes = do
       let currency = searchTry.currency
       guid <- generateGUID
       now <- getCurrentTime
       let dpRes = dpwRes.driverPoolResult
       driverStats <- runInReplica $ QDriverStats.findById dpRes.driverId
       tripQuoteDetail <- HashMap.lookup dpRes.serviceTier tripQuoteDetailsHashMap & fromMaybeM (VehicleServiceTierNotFound $ show dpRes.serviceTier)
       parallelSearchRequestCount <- Just <$> SDP.getValidSearchRequestCount searchReq.providerId dpRes.driverId now
+      farePolicy <- getFarePolicyByEstOrQuoteId searchReq.merchantOperatingCityId tripQuoteDetail.tripCategory dpRes.serviceTier searchReq.area searchTry.estimateId Nothing (Just (TransactionId (Id searchReq.transactionId)))
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/SendSearchRequestToDrivers.hs:146]: Move it to case match and also create a function which will return baseFare

**Review by hkmangla — CHANGES_REQUESTED** [review]

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs [hunk 13]
```diff
@@ -429,13 +430,19 @@ endRide handle@ServiceHandle {..} rideId req = withLogTag ("rideId-" <> rideId.g
...
 
-recalculateFareForDistance :: (MonadThrow m, Log m, MonadTime m, MonadGuid m) => ServiceHandle m -> SRB.Booking -> DRide.Ride -> Meters -> DTConf.TransporterConfig -> m (Meters, HighPrecMoney, Maybe FareParameters)
+recalculateFareForDistance :: (MonadThrow m, Log m, MonadTime m, MonadGuid m, EsqDBFlow m r, CacheFlow m r) => ServiceHandle m -> SRB.Booking -> DRide.Ride -> Meters -> DTConf.TransporterConfig -> m (Meters, HighPrecMoney, Maybe FareParameters)
 recalculateFareForDistance ServiceHandle {..} booking ride recalcDistance thresholdConfig = do
   let merchantId = booking.providerId
       oldDistance = fromMaybe 0 booking.estimatedDistance -- TODO: Fix later with rentals
 
   -- maybe compare only distance fare?
   let estimatedFare = Fare.fareSum booking.fareParams
+  vehicleAge <-
```
[the remark this answers is not in the document]
Comment (0utkarsh) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs:440]: We fetch `RiderDetails`, not `RideDetails`

File: Backend/dev/migrations/rider-app/1281-ambulance.sql [hunk 14]
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE atlas_app.booking ADD COLUMN vehicle_age int;
+
+-- Corresponding equivalent query to be run in master
+INSERT INTO atlas_app.beckn_config(
```
[the remark this answers is not in the document]
Comment (0utkarsh) [reply Backend/dev/migrations/rider-app/1281-ambulance.sql:4]: beckn_config is per vehicle category. Since I have added new category `AMBULANCE`, added it

**Review by hkmangla — APPROVED** [review]

**Pull request merged**